File: trunk/java/picocontainer/src/test/org/picocontainer/extras/ImplementationHidingComponentAdapterFactoryTestCase.java
/*****************************************************************************
 * Copyright (C) OldPicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the LICENSE.txt file.                                                     *
 *                                                                           *
 * Idea by Rachel Davies, Original code by Aslak Hellesoy and Paul Hammant   *
 *****************************************************************************/

package org.picocontainer.extras;

import org.picocontainer.PicoInitializationException;
import org.picocontainer.PicoIntrospectionException;
import org.picocontainer.tck.AbstractComponentAdapterFactoryTestCase;
import org.picocontainer.defaults.*;

import java.util.ArrayList;
import java.util.List;

public class ImplementationHidingComponentAdapterFactoryTestCase extends AbstractComponentAdapterFactoryTestCase {

    private static boolean addCalled = false;

    public static class OneConstructorArrayList extends ArrayList {
        public OneConstructorArrayList() {
            super();
        }

        public boolean add(Object o) {
            addCalled = true;
            return super.add(o);
        }
    }

    public void testCreatedComponentAdapterCreatesInstancesWhereImplementationIsHidden()
            throws NoSuchMethodException, PicoInitializationException, PicoIntrospectionException, AssignabilityRegistrationException, NotConcreteRegistrationException {
        ComponentAdapter componentAdapter = createComponentAdapterFactory().createComponentAdapter(List.class, OneConstructorArrayList.class, null);
        Object o = componentAdapter.getComponentInstance(picoContainer);
        assertTrue(o instanceof List);
        assertFalse(o instanceof OneConstructorArrayList);
        ((List) o).add("hello");
        assertTrue("Add was called", addCalled);
    }

    protected ComponentAdapterFactory createComponentAdapterFactory() {
        return new ImplementationHidingComponentAdapterFactory(new DefaultComponentAdapterFactory());
    }

}

File: dotnet/picocontainer/src/Test/Extras/ImplementationHidingComponentAdapterFactoryTestCase.cs
/*****************************************************************************
 * Copyright (C) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the license.html file.                                                    *
 *                                                                           *
 * Idea by Rachel Davies, Original code by Aslak Hellesoy and Paul Hammant   *
 * C# port by Maarten Grootendorst                                           *
 *****************************************************************************/

using System;
using System.Threading;
using System.Collections;
using csUnit;

using PicoContainer.Extras;
using PicoContainer.Lifecycle;
using PicoContainer.Defaults;
using PicoContainer.Tests.TestModel;
using PicoContainer.Tests.Tck;

namespace PicoContainer.Tests.Extras {
  /// <summary>
  /// Summary description for ImplementationHidingComponentAdapterFactoryTestCase.
  /// </summary>
  public class ReferencesSwappables {
    public Swappable swappable;

    public ReferencesSwappables(Swappable swappable) {
      this.swappable = swappable;
    }
  }

  public interface Swappable {
    String getCheese();
  }

  public class ConcreteSwappable : Swappable {
    public String getCheese() {
      return "Edam";
    }
  }

  public class ImplementationHidingComponentAdapterFactoryTestCase : AbstractComponentAdapterFactoryTestCase {



    public void testCreatedComponentAdapterCreatesInstancesWhereImplementationIsHidden() {
      ComponentAdapter componentAdapter = CreateComponentAdapterFactory().CreateComponentAdapter(typeof(Swappable), typeof(ConcreteSwappable), null);
      Swappable swappable = (Swappable) componentAdapter.GetComponentInstance(picoContainer);
      Assert.False(swappable is ConcreteSwappable);
      Assert.Equals("Edam", swappable.getCheese());
    }

    public void testHotSwap() {
      ImplementationHidingComponentAdapterFactory componentAdapterFactory = (ImplementationHidingComponentAdapterFactory) CreateComponentAdapterFactory();
      ComponentAdapter componentAdapter = componentAdapterFactory.CreateComponentAdapter(typeof(Swappable), typeof(ConcreteSwappable), null);
      Swappable swappable = (Swappable) componentAdapter.GetComponentInstance(picoContainer);
      Assert.False(swappable is ConcreteSwappable);
      Assert.Equals("Edam", swappable.getCheese());

      //TODO - what is this last line for, throws something on errror or should have an Assert. following ?
      componentAdapterFactory.HotSwap(typeof(Swappable));
    }


    protected override ComponentAdapterFactory CreateComponentAdapterFactory() {
      return new ImplementationHidingComponentAdapterFactory(new DefaultComponentAdapterFactory());
    }

    public static void Main() {

      Type c = typeof(ConcreteSwappable);
      
      ImplementationHidingComponentAdapterFactoryTestCase d = new ImplementationHidingComponentAdapterFactoryTestCase ();
      d.SetUp();
      d.testHotSwap();
    }

  }
}
